File: python/day15.py
from intcode import *
from copy import deepcopy

def solve(c, part):
    seen = set()
    dirs = [-1j, 1j, -1, 1]
    pos = 0j
    steps = 0
    queue = deque([(steps, pos, c)])
    while queue:
        steps, pos, c = queue.popleft()
        c.run()
        for i in range(4):
            npos = pos + dirs[i]
            if npos not in seen:
                nsteps = steps+1
                seen.add(npos)
                nc = deepcopy(c)
                nc.receive(i+1)
                nc.run()
                status = nc.get_from_out_queue()
                if status == 1:
                    queue.append((nsteps, npos, nc))
                elif status == 2 and part == 1:
                    return nsteps, nc
    return steps


c = Computer(read_intcode_input(15))
steps, comp = solve(c, 1)
flood = solve(comp, 2)

print(steps)
print(flood)

File: clojure/day15.clj
(ns day15
  (:require [intcode :as ic]
            aoc))

(def directions
  {1 [0 -1]
   2 [0 1]
   3 [-1 0]
   4 [1 0]})

(defn- run [computer]
  (loop [queue (conj aoc/empty-queue [0 [0 0] 1 computer])
         seen #{}]
    (let [[steps pos status computer] (peek queue)]
      (if (= 2 status) [steps computer] ; part 1
          (let [seen' (conj seen pos)
                queue' (reduce (fn [q [i dir]]
                                 (let [pos' (aoc/pt+ pos dir)
                                       comp' (ic/in-run-out computer i)
                                       out (:output comp')]
                                   (if (or (seen pos') (zero? out)) q
                                       (conj q [(inc steps) pos' out comp']))))
                               (pop queue)
                               directions)]
            (if (empty? queue')
              steps ; part 2
              (recur queue' seen')))))))

(defn solve [filename]
  (let [computer (ic/initialize-from-file filename)
        [pt1 comp'] (run computer)]
    [pt1 (run comp')]))


(solve 15)
